core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,163 @@
+package artifacts
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	atomicfile "github.com/natefinch/atomic"
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type FileCache struct {
+	root string
+}
+
+func NewFileCache() *FileCache {
+	dir, found := os.LookupEnv("WANDB_CACHE_DIR")
+	if !found {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			return nil
+		}
+		dir = filepath.Join(userCacheDir, "wandb")
+	}
+	return &FileCache{root: filepath.Join(dir, "artifacts")}
+}
+
+// Copy a file into the cache and return the B64MD5 cache key.
+func (c *FileCache) AddFile(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return c.Write(f)
+}
+
+// This is a convenience function intended for running in the background; instead of
+// returning an error it just logs it. This is often correct, since failing to add a
+// file to the cache shouldn't be a critical error.
+//
+// If a digest is passed in it will be checked against the calculated hash.
+func (c *FileCache) AddFileAndCheckDigest(path string, digest *string) {
```

Comment:
Using a nil pointer to indicate a missing value is a little scary. Can the digest ever be the empty string, and if not, then could the check be changed to

```golang
if digest != "" && digest != b64md5 {
```